src/v/ssx/rwlock.h

```diff
@@ -0,0 +1,169 @@
+        logging_rwlock* _lock;
+        ss::rwlock::holder _h;
+        holder_id _id;
+    };
+
+    explicit logging_rwlock(ss::logger& logger)
+      : _logger(logger) {}
+
+    explicit logging_rwlock(ss::logger& logger, resource_str_func fn)
+      : _logger(logger)
+      , _resource_str_gen(std::move(fn)) {}
+
+    ss::future<holder> hold_read_lock(
+      ss::semaphore::time_point take_deadline
+      = ss::semaphore::time_point::max(),
+      Clock::duration log_timeout = default_log_timeout) {
+        return lock_with_log(
+          [this, take_deadline]() {
+              return _lock.hold_read_lock(take_deadline);
+          },
+          "Read",
+          log_timeout);
+    }
+
+    ss::future<holder> hold_write_lock(
+      ss::semaphore::time_point take_deadline
+      = ss::semaphore::time_point::max(),
+      Clock::duration log_timeout = default_log_timeout) {
+        return lock_with_log(
+          [this, take_deadline]() {
+              return _lock.hold_write_lock(take_deadline);
+          },
+          "Write",
+          log_timeout);
+    }
+
+    bool locked() const { return _lock.locked(); }
+
+    bool try_write_lock() {
+        auto ret = _lock.try_write_lock();
+        if (ret) {
+            vlog(
+              _logger.trace,
+              "Write lock on {} acquired immediately",
+              _resource_str_gen());
+        }
+        return ret;
+    }
+
+    void write_unlock() {
+        vlog(_logger.trace, "Dropping write lock on {}", _resource_str_gen());
+        _lock.write_unlock();
+    }
+
+private:
+    friend struct holder;
+
+    template<typename F>
+    requires ReturnsFutureHolder<F> ss::future<holder> lock_with_log(
+      F lock_func, const char* acquire_type, Clock::duration log_timeout) {
+        auto id = _next_id++;
+        vlog(
+          _logger.trace,
+          "{} lock on {} ({}) being acquired",
+          acquire_type,
+          _resource_str_gen(),
+          id);
+        ss::timer<Clock> log_timer;
+        const auto start_time = ss::lowres_clock::now();
+        log_timer.set_callback([id, start_time, acquire_type, this] {
+            vlog(
+              _logger.error,
+              "{} lock on {} ({}) has taken {}ns so far",
+              acquire_type,
+              _resource_str_gen(),
+              id,
+              (ss::lowres_clock::now() - start_time).count());
+        });
+        log_timer.arm(log_timeout);
+        auto stop_timer = ss::defer([&log_timer] { log_timer.cancel(); });
+        auto h = co_await lock_func();
+        auto acquire_duration = ss::lowres_clock::now() - start_time;
+        vlog(
+          _logger.trace,
+          "{} lock on {} ({}) acquired after {}ns",
+          acquire_type,
+          _resource_str_gen(),
+          id,
+          acquire_duration.count());
+        co_return holder(this, std::move(h), id);
+    }
+
+    ss::logger& _logger;
+    ss::rwlock _lock;
+    const resource_str_func _resource_str_gen;
```

Comment:
if you are defining them on the ctor, why not forrce this to be a `std::string_view`  - similara to honey badger, codepoints are always statically defined or codegenerated.

Reply:
Yeah I'd considered making this something like a `__FILE__, __LINE__`, or `const char*` to discourage references to strings but opted for this to be a bit more expressive: my immediate use case needs this to be at runtime to log the filename of a segment which is currently only exposed as a string copy from segments (not opposed to making this a ref though).